src/MCMCMonitorDataManager/MCMCMonitorData.ts

```diff
@@ -334,4 +333,18 @@ const computeEffectiveWarmupIterations = (data: MCMCMonitorData, requested: numb
     } else {
         return requested
     }
-}
\ No newline at end of file
+}
+
+
+const invalidateStats = <T extends SequenceStatsDict | VariableStatsDict>(statsDict: T, key?: string): T => {
```

Comment:
So there's a few differences here.

I've written this function to handle invalidating either a single `SequenceStats`/`VariableStats` object, or the entire dictionary of such objects, as needed. (That's why it's called differently on l. 191-192, where we only touch one record, vs l. 255-256, which invalidate the whole dictionary.)

This proposal would add an `isUpToDate` key on the dictionary itself, rather than on the elements of the dictionary (individual sets of stats).

I can imagine a method that applies to the individual stats objects, like:
```Typescript
const invalidateOneSetOfStats = (stats: SequenceStats | VariableStats) => {
    return {
        ...stats,
        isUpToDate: false
    }
}
```
However the type inference for the return value is sloppier: VSCode annotates this return value as being a union of either `(the full definition of SequenceStats)` or `(the full definition of VariableStats)` without being able to recover the actual type, and then you get duck typing everywhere after that in the code, which is at least harder to follow. If I do:

```Typescript
const invalidateOneSetOfStatsTyped = <T extends SequenceStats | VariableStats>(stats: T): T => {
    return {
        ...stats,
        isUpToDate: false
    }
}
```
with the explicit type parameter, then when I call it (without specifying type) it correctly recognizes that the return value is the same as the parameter value, so the annotations continue to refer to `SequenceStats`/`VariableStats` as appropriate.

The other thing is that the implementation I used will edit the object in-place, vs doing a memory copy. Not likely to be an issue performance-wise, but it does feel better to avoid the copy. Anyway I'm not really sure how I'd do the destructuring-overwrite construct on the whole dictionary--I could probably figure it out, but I suspect it would be a lot less idiomatic.

Reply:
Okay I see, it's trickier than I had thought at first glance.
Regarding editing object in-place, I think generally it's better to make a copy to avoid unintended side effects. Computation time is rarely an issue.